File: server/services/items.js
'use strict';
/** المجموعات والأصناف. */
const db = require('../db');
const V = require('../lib/validate');
const { uuid } = require('../lib/ids');
const M = require('../lib/money');

// ------------------------------------------------------------- المجموعات
function mapCategory(row) {
  if (!row) return null;
  return {
    id: row.id,
    code: row.code,
    name: row.name,
    parent_id: row.parent_id,
    parent_name: row.parent_name === undefined ? undefined : row.parent_name,
    description: row.description,
    items_count: row.items_count,
    created_at: row.created_at,
  };
}

function listCategories() {
  return db.all(
    `SELECT c.*, p.name AS parent_name,
        (SELECT COUNT(*) FROM items i WHERE i.category_id = c.id) AS items_count
      FROM item_categories c LEFT JOIN item_categories p ON p.id = c.parent_id
      ORDER BY c.name`,
  ).map(mapCategory);
}

function createCategory(payload, actor) {
  const code = V.str(payload.code, 'كود المجموعة', { required: true, max: 30 });
  const name = V.str(payload.name, 'اسم المجموعة', { required: true, max: 150 });
  const parentId = V.str(payload.parent_id, 'المجموعة الرئيسية', { max: 40 }) || null;
  if (db.get('SELECT id FROM item_categories WHERE code = :c', { c: code })) throw V.conflict('كود المجموعة مستخدم مسبقاً');
  if (parentId && !db.get('SELECT id FROM item_categories WHERE id = :id', { id: parentId })) throw V.bad('المجموعة الرئيسية غير موجودة');
  const id = uuid();
  db.run(
    `INSERT INTO item_categories (id, code, name, parent_id, description, created_at)
     VALUES (:id, :code, :name, :parent_id, :description, :created_at)`,
    { id, code, name, parent_id: parentId, description: V.str(payload.description, 'الوصف', { max: 500 }), created_at: db.nowIso() },
  );
  db.audit({ user: actor, action: 'CATEGORY_CREATE', entityType: 'item_category', entityId: id, details: { code, name } });
  return mapCategory(db.get('SELECT * FROM item_categories WHERE id = :id', { id }));
}

function updateCategory(id, payload, actor) {
  const current = db.get('SELECT * FROM item_categories WHERE id = :id', { id });
  if (!current) throw V.notFound('المجموعة غير موجودة');
  const code = V.str(payload.code === undefined ? current.code : payload.code, 'كود المجموعة', { required: true, max: 30 });
  const name = V.str(payload.name === undefined ? current.name : payload.name, 'اسم المجموعة', { required: true, max: 150 });
  let parentId = payload.parent_id === undefined ? current.parent_id : (V.str(payload.parent_id, 'المجموعة الرئيسية', { max: 40 }) || null);
  if (parentId === id) parentId = null;
  if (code !== current.code && db.get('SELECT id FROM item_categories WHERE code = :c AND id <> :id', { c: code, id })) {
    throw V.conflict('كود المجموعة مستخدم مسبقاً');
  }
  db.run(
    'UPDATE item_categories SET code = :code, name = :name, parent_id = :parent_id, description = :description WHERE id = :id',
    { id, code, name, parent_id: parentId, description: V.str(payload.description === undefined ? current.description : payload.description, 'الوصف', { max: 500 }) },
  );
  db.audit({ user: actor, action: 'CATEGORY_UPDATE', entityType: 'item_category', entityId: id, details: { code } });
  return mapCategory(db.get('SELECT * FROM item_categories WHERE id = :id', { id }));
}

function removeCategory(id, actor) {
  const current = db.get('SELECT * FROM item_categories WHERE id = :id', { id });
  if (!current) throw V.notFound('المجموعة غير موجودة');
  const count = db.pluck('SELECT COUNT(*) AS c FROM items WHERE category_id = :id', { id });
  if (count > 0) throw V.conflict(`لا يمكن حذف المجموعة لاحتوائها على ${count} صنف`);
  db.run('DELETE FROM item_categories WHERE id = :id', { id });
  db.audit({ user: actor, action: 'CATEGORY_DELETE', entityType: 'item_category', entityId: id, details: { code: current.code } });
  return { ok: true };
}

// ---------------------------------------------------------------- الأصناف
function mapItem(row) {
  if (!row) return null;
  return {
    id: row.id,
    item_code: row.item_code,
    category_id: row.category_id,
    category_name: row.category_name === undefined ? undefined : row.category_name,
    name_ar: row.name_ar,
    name_en: row.name_en,
    barcode: row.barcode,
    unit: row.unit,
    cost_price: M.toMajor(row.cost_price),
    sale_price: M.toMajor(row.sale_price),
    tax_rate: row.tax_rate,
    is_active: !!row.is_active,
    notes: row.notes,
    created_at: row.created_at,
    updated_at: row.updated_at,
  };
}

function listItems({
  search = '', categoryId = '', activeOnly = false, status = '',
  minPrice = '', maxPrice = '', taxRate = '', limit = 0,
} = {}) {
  let activeVal = -1;
  if (activeOnly || status === 'active') activeVal = 1;
  else if (status === 'inactive') activeVal = 0;

  const minP = minPrice !== '' && minPrice !== undefined ? M.toMinor(Number(minPrice)) : -1;
  const maxP = maxPrice !== '' && maxPrice !== undefined ? M.toMinor(Number(maxPrice)) : -1;
  const tax = taxRate !== '' && taxRate !== undefined ? Number(taxRate) : -1;

  const rows = db.all(
    `SELECT i.*, c.name AS category_name
      FROM items i LEFT JOIN item_categories c ON c.id = i.category_id
      WHERE (:active = -1 OR i.is_active = :active)
        AND (:cat = '' OR i.category_id = :cat)
        AND (:min_p < 0 OR i.sale_price >= :min_p)
        AND (:max_p < 0 OR i.sale_price <= :max_p)
        AND (:tax < 0 OR i.tax_rate = :tax)
        AND (:q = '' OR i.name_ar LIKE :like OR i.name_en LIKE :like OR i.item_code LIKE :like OR i.barcode LIKE :like)
      ORDER BY i.name_ar
      LIMIT CASE WHEN :lim > 0 THEN :lim ELSE -1 END`,
    {
      active: activeVal,
      cat: categoryId || '',
      min_p: minP,
      max_p: maxP,
      tax,
      q: search,
      like: `%${search}%`,
      lim: limit || 0,
    },
  );
  return rows.map(mapItem);
}

function getItemRaw(id) {
  const row = db.get('SELECT * FROM items WHERE id = :id', { id });
  if (!row) throw V.notFound('الصنف غير موجود');
  return row;
}

function getItem(id) {
  return mapItem(db.get('SELECT i.*, c.name AS category_name FROM items i LEFT JOIN item_categories c ON c.id = i.category_id WHERE i.id = :id', { id }));
}

function nextItemCode() {
  const rows = db.all("SELECT item_code FROM items WHERE item_code LIKE 'IT-%'");
  let max = 0;
  for (const r of rows) {
    const n = Number(String(r.item_code).replace(/^IT-/, ''));
    if (Number.isFinite(n) && n > max) max = n;
  }
  return `IT-${String(max + 1).padStart(4, '0')}`;
}

function validateItem(payload, { isNew }) {
  const out = {
    item_code: V.str(payload.item_code, 'كود الصنف', { max: 40 }) || (isNew ? nextItemCode() : ''),
    category_id: V.str(payload.category_id, 'المجموعة', { max: 40 }) || null,
    name_ar: V.str(payload.name_ar, 'اسم الصنف', { required: isNew, max: 250 }),
    name_en: V.str(payload.name_en, 'الاسم بالإنجليزية', { max: 250 }),
    barcode: V.str(payload.barcode, 'الباركود', { max: 60 }),
    unit: V.str(payload.unit, 'وحدة القياس', { max: 40 }) || 'حبة',
    cost_price: M.toMinor(V.num(payload.cost_price, 'سعر التكلفة', { min: 0, max: 1e9, def: 0 })),
    sale_price: M.toMinor(V.num(payload.sale_price, 'سعر البيع', { min: 0, max: 1e9, def: 0 })),
    tax_rate: V.num(payload.tax_rate, 'نسبة الضريبة', { min: 0, max: 100, def: 15 }),
    is_active: V.bool(payload.is_active, true) ? 1 : 0,
    notes: V.str(payload.notes, 'الملاحظات', { max: 1000 }),
  };
  if (out.category_id && !db.get('SELECT id FROM item_categories WHERE id = :id', { id: out.category_id })) {
    throw V.bad('المجموعة المحددة غير موجودة');
  }
  return out;
}

function createItem(payload, actor) {
  const data = validateItem(payload, { isNew: true });
  if (db.get('SELECT id FROM items WHERE item_code = :c', { c: data.item_code })) throw V.conflict('كود الصنف مستخدم مسبقاً');
  const id = uuid();
  const now = db.nowIso();
  db.run(
    `INSERT INTO items (id, item_code, category_id, name_ar, name_en, barcode, unit, cost_price, sale_price,
        tax_rate, is_active, notes, created_at, updated_at)
     VALUES (:id, :item_code, :category_id, :name_ar, :name_en, :barcode, :unit, :cost_price, :sale_price,
        :tax_rate, :is_active, :notes, :created_at, :updated_at)`,
    { ...data, id, created_at: now, updated_at: now },
  );
  db.audit({ user: actor, action: 'ITEM_CREATE', entityType: 'item', entityId: id, details: { code: data.item_code, name: data.name_ar } });
  return getItem(id);
}

function updateItem(id, payload, actor) {
  const current = getItemRaw(id);
  const merged = {
    ...current,
    cost_price: M.toMajor(current.cost_price),
    sale_price: M.toMajor(current.sale_price),
    ...payload,
  };
  const data = validateItem(merged, { isNew: false });
  if (!data.item_code) data.item_code = current.item_code;
  if (data.item_code !== current.item_code && db.get('SELECT id FROM items WHERE item_code = :c AND id <> :id', { c: data.item_code, id })) {
    throw V.conflict('كود الصنف مستخدم مسبقاً');
  }
  db.run(
    `UPDATE items SET item_code = :item_code, category_id = :category_id, name_ar = :name_ar, name_en = :name_en,
        barcode = :barcode, unit = :unit, cost_price = :cost_price, sale_price = :sale_price, tax_rate = :tax_rate,
        is_active = :is_active, notes = :notes, updated_at = :updated_at
      WHERE id = :id`,
    { ...data, id, updated_at: db.nowIso() },
  );
  db.audit({ user: actor, action: 'ITEM_UPDATE', entityType: 'item', entityId: id, details: { code: data.item_code } });
  return getItem(id);
}

function removeItem(id, actor) {
  const current = getItemRaw(id);
  db.run('DELETE FROM items WHERE id = :id', { id });
  db.audit({ user: actor, action: 'ITEM_DELETE', entityType: 'item', entityId: id, details: { code: current.item_code } });
  return { ok: true };
}

module.exports = {
  listCategories, createCategory, updateCategory, removeCategory,
  listItems, getItem, getItemRaw, createItem, updateItem, removeItem, mapItem, nextItemCode,
};

File: internal/services/items.go
package services

import (
	"database/sql"
	"errors"
	"fmt"
	"strings"

	"raseen/internal/crypto"
	"raseen/internal/db"
	"raseen/internal/models"
)

type ItemService struct {
	db *db.DB
}

func NewItemService(d *db.DB) *ItemService {
	return &ItemService{db: d}
}

func (s *ItemService) ListCategories() ([]models.ItemCategory, error) {
	rows, err := s.db.Query(`
		SELECT id, code, name, parent_id, description, created_at
		FROM item_categories ORDER BY name ASC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	list := make([]models.ItemCategory, 0)
	for rows.Next() {
		var cat models.ItemCategory
		var parent sql.NullString
		if err := rows.Scan(&cat.ID, &cat.Code, &cat.Name, &parent, &cat.Description, &cat.CreatedAt); err == nil {
			if parent.Valid {
				cat.ParentID = &parent.String
			}
			list = append(list, cat)
		}
	}
	return list, nil
}

func (s *ItemService) CreateCategory(c *models.ItemCategory) error {
	if c.Name == "" {
		return errors.New("اسم المجموعة مطلوب")
	}
	if c.ID == "" {
		c.ID = crypto.UUID()
	}
	if c.Code == "" {
		var count int
		_ = s.db.QueryRow("SELECT COUNT(*) + 1 FROM item_categories").Scan(&count)
		c.Code = fmt.Sprintf("CAT-%03d", count)
	}
	c.CreatedAt = db.NowIso()

	_, err := s.db.Exec(`
		INSERT INTO item_categories (id, code, name, parent_id, description, created_at)
		VALUES (?, ?, ?, ?, ?, ?)
	`, c.ID, c.Code, c.Name, c.ParentID, c.Description, c.CreatedAt)
	return err
}

type ItemListItem struct {
	models.Item
	CostPriceMajor float64 `json:"cost_price"`
	SalePriceMajor float64 `json:"sale_price"`
	CategoryName   string  `json:"category_name"`
}

func (s *ItemService) ListItems(search, categoryID string, activeOnly bool) ([]ItemListItem, error) {
	query := `
		SELECT i.id, i.item_code, i.category_id, i.name_ar, i.name_en, i.barcode,
		       i.unit, i.cost_price, i.sale_price, i.tax_rate, i.is_active,
		       i.notes, i.created_at, i.updated_at, COALESCE(c.name, '—') AS category_name
		FROM items i
		LEFT JOIN item_categories c ON c.id = i.category_id
		WHERE (? = 0 OR i.is_active = 1)
		  AND (? = '' OR i.category_id = ?)
		  AND (? = '' OR i.name_ar LIKE ? OR i.name_en LIKE ? OR i.item_code LIKE ? OR i.barcode LIKE ?)
		ORDER BY i.name_ar ASC
	`
	activeParam := 0
	if activeOnly {
		activeParam = 1
	}
	like := "%" + search + "%"

	rows, err := s.db.Query(query, activeParam, categoryID, categoryID, search, like, like, like, like)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	list := make([]ItemListItem, 0)
	for rows.Next() {
		var item ItemListItem
		var catID sql.NullString
		if err := rows.Scan(
			&item.ID, &item.ItemCode, &catID, &item.NameAr, &item.NameEn, &item.Barcode,
			&item.Unit, &item.CostPrice, &item.SalePrice, &item.TaxRate, &item.IsActive,
			&item.Notes, &item.CreatedAt, &item.UpdatedAt, &item.CategoryName,
		); err == nil {
			if catID.Valid {
				item.CategoryID = &catID.String
			}
			item.CostPriceMajor = models.ToMajor(item.CostPrice)
			item.SalePriceMajor = models.ToMajor(item.SalePrice)
			list = append(list, item)
		}
	}
	return list, nil
}

func (s *ItemService) GetItem(id string) (*ItemListItem, error) {
	var item ItemListItem
	var catID sql.NullString

	err := s.db.QueryRow(`
		SELECT i.id, i.item_code, i.category_id, i.name_ar, i.name_en, i.barcode,
		       i.unit, i.cost_price, i.sale_price, i.tax_rate, i.is_active,
		       i.notes, i.created_at, i.updated_at, COALESCE(c.name, '—')
		FROM items i
		LEFT JOIN item_categories c ON c.id = i.category_id
		WHERE i.id = ?
	`, id).Scan(
		&item.ID, &item.ItemCode, &catID, &item.NameAr, &item.NameEn, &item.Barcode,
		&item.Unit, &item.CostPrice, &item.SalePrice, &item.TaxRate, &item.IsActive,
		&item.Notes, &item.CreatedAt, &item.UpdatedAt, &item.CategoryName,
	)
	if err != nil {
		return nil, err
	}
	if catID.Valid {
		item.CategoryID = &catID.String
	}
	item.CostPriceMajor = models.ToMajor(item.CostPrice)
	item.SalePriceMajor = models.ToMajor(item.SalePrice)
	return &item, nil
}

func (s *ItemService) nextItemCode() string {
	var maxNum int
	rows, err := s.db.Query("SELECT item_code FROM items WHERE item_code LIKE 'ITM-%'")
	if err == nil {
		defer rows.Close()
		for rows.Next() {
			var code string
			if err := rows.Scan(&code); err == nil {
				var n int
				if _, err := fmt.Sscanf(code, "ITM-%d", &n); err == nil && n > maxNum {
					maxNum = n
				}
			}
		}
	}
	return fmt.Sprintf("ITM-%04d", maxNum+1)
}

func (s *ItemService) CreateItem(item *models.Item) error {
	if item.NameAr == "" {
		return errors.New("اسم الصنف بالعربية مطلوب")
	}
	if item.ID == "" {
		item.ID = crypto.UUID()
	}
	if item.ItemCode == "" {
		item.ItemCode = s.nextItemCode()
	}
	if item.Unit == "" {
		item.Unit = "حبة"
	}
	if item.TaxRate <= 0 {
		item.TaxRate = 15.0
	}

	now := db.NowIso()
	item.CreatedAt = now
	item.UpdatedAt = now
	item.IsActive = 1

	_, err := s.db.Exec(`
		INSERT INTO items (
			id, item_code, category_id, name_ar, name_en, barcode,
			unit, cost_price, sale_price, tax_rate, is_active, notes,
			created_at, updated_at
		) VALUES (
			?, ?, ?, ?, ?, ?,
			?, ?, ?, ?, ?, ?,
			?, ?
		)
	`,
		item.ID, item.ItemCode, item.CategoryID, item.NameAr, item.NameEn, item.Barcode,
		item.Unit, item.CostPrice, item.SalePrice, item.TaxRate, item.IsActive, item.Notes,
		item.CreatedAt, item.UpdatedAt,
	)
	return err
}

func (s *ItemService) UpdateItem(id string, item *models.Item) error {
	now := db.NowIso()
	item.UpdatedAt = now

	_, err := s.db.Exec(`
		UPDATE items SET
			category_id = ?, name_ar = ?, name_en = ?, barcode = ?,
			unit = ?, cost_price = ?, sale_price = ?, tax_rate = ?,
			is_active = ?, notes = ?, updated_at = ?
		WHERE id = ?
	`,
		item.CategoryID, item.NameAr, item.NameEn, item.Barcode,
		item.Unit, item.CostPrice, item.SalePrice, item.TaxRate,
		item.IsActive, item.Notes, now, id,
	)
	return err
}

func (s *ItemService) DeleteItem(id string) error {
	var count int
	_ = s.db.QueryRow("SELECT COUNT(*) FROM invoice_items WHERE item_id = ?", id).Scan(&count)
	if count > 0 {
		return errors.New("لا يمكن حذف الصنف لوجود حركات مبيعات وفواتير مسجلة به")
	}
	_, err := s.db.Exec("DELETE FROM items WHERE id = ?", id)
	return err
}

type ImportItemInput struct {
	ItemCode  string  `json:"item_code"`
	NameAr    string  `json:"name_ar"`
	NameEn    string  `json:"name_en"`
	Category  string  `json:"category"`
	Barcode   string  `json:"barcode"`
	Unit      string  `json:"unit"`
	CostPrice float64 `json:"cost_price"`
	SalePrice float64 `json:"sale_price"`
	TaxRate   float64 `json:"tax_rate"`
	Notes     string  `json:"notes"`
}

type ImportItemsResult struct {
	Total     int `json:"total"`
	Created   int `json:"created"`
	Updated   int `json:"updated"`
	Errors    int `json:"errors"`
}

func (s *ItemService) BatchImportItems(items []ImportItemInput) (*ImportItemsResult, error) {
	if len(items) == 0 {
		return &ImportItemsResult{}, nil
	}

	// Cache or create categories
	cats, err := s.ListCategories()
	if err != nil {
		return nil, err
	}
	catMap := make(map[string]string) // name -> id
	for _, c := range cats {
		catMap[strings.ToLower(strings.TrimSpace(c.Name))] = c.ID
	}

	tx, err := s.db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	stmtFindByName, err := tx.Prepare("SELECT id FROM items WHERE LOWER(TRIM(name_ar)) = LOWER(TRIM(?)) LIMIT 1")
	if err != nil {
		return nil, err
	}
	defer stmtFindByName.Close()

	stmtFindByCode, err := tx.Prepare("SELECT id FROM items WHERE item_code = ? AND item_code != '' LIMIT 1")
	if err != nil {
		return nil, err
	}
	defer stmtFindByCode.Close()

	stmtInsertCat, err := tx.Prepare("INSERT INTO item_categories (id, code, name, description, created_at) VALUES (?, ?, ?, '', ?)")
	if err != nil {
		return nil, err
	}
	defer stmtInsertCat.Close()

	stmtInsertItem, err := tx.Prepare(`
		INSERT INTO items (
			id, item_code, category_id, name_ar, name_en, barcode,
			unit, cost_price, sale_price, tax_rate, is_active, notes,
			created_at, updated_at
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 1, ?, ?, ?)
	`)
	if err != nil {
		return nil, err
	}
	defer stmtInsertItem.Close()

	stmtUpdateItem, err := tx.Prepare(`
		UPDATE items SET
			item_code = CASE WHEN ? != '' THEN ? ELSE item_code END,
			category_id = COALESCE(?, category_id),
			name_en = CASE WHEN ? != '' THEN ? ELSE name_en END,
			barcode = CASE WHEN ? != '' THEN ? ELSE barcode END,
			unit = CASE WHEN ? != '' THEN ? ELSE unit END,
			cost_price = ?,
			sale_price = ?,
			tax_rate = ?,
			notes = CASE WHEN ? != '' THEN ? ELSE notes END,
			updated_at = ?
		WHERE id = ?
	`)
	if err != nil {
		return nil, err
	}
	defer stmtUpdateItem.Close()

	now := db.NowIso()
	res := &ImportItemsResult{Total: len(items)}

	for _, itm := range items {
		nameAr := strings.TrimSpace(itm.NameAr)
		if nameAr == "" {
			res.Errors++
			continue
		}

		var catID *string
		if itm.Category != "" {
			catKey := strings.ToLower(strings.TrimSpace(itm.Category))
			if cid, exists := catMap[catKey]; exists {
				catID = &cid
			} else {
				newCid := crypto.UUID()
				newCode := fmt.Sprintf("CAT-%03d", len(catMap)+1)
				if _, err := stmtInsertCat.Exec(newCid, newCode, itm.Category, now); err == nil {
					catMap[catKey] = newCid
					catID = &newCid
				}
			}
		}

		// Check if exists by code or by Arabic name
		var existingID string
		if itm.ItemCode != "" {
			_ = stmtFindByCode.QueryRow(itm.ItemCode).Scan(&existingID)
		}
		if existingID == "" {
			_ = stmtFindByName.QueryRow(nameAr).Scan(&existingID)
		}

		unit := itm.Unit
		if unit == "" {
			unit = "حبة"
		}
		taxRate := itm.TaxRate
		if !validAmount(taxRate) || taxRate > 100 || !validAmount(itm.CostPrice) || !validAmount(itm.SalePrice) { return nil,errors.New("سعر أو ضريبة الصنف غير صالح") }
		costMinor := models.ToMinor(itm.CostPrice)
		saleMinor := models.ToMinor(itm.SalePrice)

		if existingID != "" {
			// Update
			_, err := stmtUpdateItem.Exec(
				itm.ItemCode, itm.ItemCode,
				catID,
				itm.NameEn, itm.NameEn,
				itm.Barcode, itm.Barcode,
				unit, unit,
				costMinor, saleMinor, taxRate,
				itm.Notes, itm.Notes,
				now, existingID,
			)
			if err != nil {
				res.Errors++
			} else {
				res.Updated++
			}
		} else {
			// Insert
			itemCode := itm.ItemCode
			if itemCode == "" {
				itemCode = "ITM-" + crypto.UUID()
			}
			newID := crypto.UUID()
			_, err := stmtInsertItem.Exec(
				newID, itemCode, catID, nameAr, itm.NameEn, itm.Barcode,
				unit, costMinor, saleMinor, taxRate, itm.Notes,
				now, now,
			)
			if err != nil {
				res.Errors++
			} else {
				res.Created++
			}
		}
	}

	if err := tx.Commit(); err != nil {
		return nil, err
	}
	return res, nil
}
